commit 3e7bc80b35b43d495a56b9cb0e9e15f58381f77b
Author: Jason Dillaman <dillaman@redhat.com>
Date:   Wed Nov 27 09:32:28 2019 -0500

    mgr: ensure new daemons are properly indexed by hostname
    
    The hostname key was never populated when storing a daemon within
    the daemon state index for the first time. This results in the
    "by_server" index being incorrect.
    
    Signed-off-by: Jason Dillaman <dillaman@redhat.com>

diff --git a/src/mgr/DaemonServer.cc b/src/mgr/DaemonServer.cc
index 6d086179c2c..7738c380679 100644
--- a/src/mgr/DaemonServer.cc
+++ b/src/mgr/DaemonServer.cc
@@ -397,6 +397,7 @@ bool DaemonServer::handle_open(const ref_t<MMgrOpen>& m)
 
   DaemonStatePtr daemon;
   if (daemon_state.exists(key)) {
+    dout(20) << "updating existing DaemonState for " << key << dendl;
     daemon = daemon_state.get(key);
   }
   if (m->service_daemon && !daemon) {
@@ -407,13 +408,17 @@ bool DaemonServer::handle_open(const ref_t<MMgrOpen>& m)
     daemon_state.insert(daemon);
   }
   if (daemon) {
-    dout(20) << "updating existing DaemonState for " << m->daemon_name << dendl;
+    if (m->service_daemon) {
+      // update the metadata through the daemon state index to
+      // ensure it's kept up-to-date
+      daemon_state.update_metadata(daemon, m->daemon_metadata);
+    }
+
     std::lock_guard l(daemon->lock);
     daemon->perf_counters.clear();
 
     daemon->service_daemon = m->service_daemon;
     if (m->service_daemon) {
-      daemon->set_metadata(m->daemon_metadata);
       daemon->service_status = m->daemon_status;
 
       utime_t now = ceph_clock_now();
